자유게시판 서비스 단위 테스트 구현

PhilSoGooood opened the pull request:
안녕하세요 리뷰어님, PhilSoHanse입니다.
바쁘신 와중에 리뷰를 봐주셔서 대단히 감사합니다.

이번에는 기존 E2E 테스트 말고 자유게시판 서비스 기능을 단위 테스트하였습니다.
테스트 방법이 아직 익숙치 않아 일단 해보았는데 이게 단위 테스트가 맞을까요?

외부 통신이 필요하지 않은 이미 구현한 레포지토리를 모킹하는 것은 불필요한 것 같아 하지 않았습니다.

감사합니다.

PhilSoGooood committed:
:recycle: 생성자 어노테이션 추가

- 테스트를 위한 전체 생성자 추가

Co-authored-by: rkolx <olx.rko.o@gmail.com>

PhilSoGooood committed:
:recycle: 전체 조회 시 삭제된 글은 제외하도록 수정

Co-authored-by: rkolx <olx.rko.o@gmail.com>

PhilSoGooood committed:
:white_check_mark: PlaygroundService 단위 테스트 구현

- create, findDetail, findList, update, delete 테스트

Co-authored-by: rkolx <olx.rko.o@gmail.com>

ku-kim commented on noriter/src/main/java/com/codewarts/noriter/article/dto/playground/PlaygroundCreateRequest.java:
```diff
@@ -6,10 +6,12 @@
 import lombok.NoArgsConstructor;
 
 @Getter
+@AllArgsConstructor
```
+a) AllArgsConstructor로 만들어진 생성자를 사용하고 있을까요?
개인적으로 AllArgsConstructor 사용은 잘 안하고있어서요. 혹시 특별히 사용한 이유가 있을까요?

https://kwonnam.pe.kr/wiki/java/lombok/pitfall

PhilSoGooood replied:
스프링은 바로 Jackson 라이브러리의 ObjectMapper를 사용하여 JSON으로 맵핑한다는 것을 알고 있어 기본 생성자만 사용하였었습니다. 그러나, 테스트 코드를 작성하면서 dto를 생성해야하는 경우가 있어 AllArgsConstructor를 추가해주었습니다!

롬복의 경우 저런 문제는 생각을 못했었네요! 다른 부분들은 Builder가 있어서 상관이 없는데, 그렇지 않은 부분들은 유의하도록 하겠습니다.

ku-kim commented on noriter/src/test/java/com/codewarts/noriter/article/unit/service/PlaygroundServiceTest.java:
```diff
@@ -0,0 +1,342 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Transactional
+@SpringBootTest
```
테스트 작성해보셨군요.! 아주 좋네요 ㅎㅎ 
시작이 반이니 금방 더 좋아지실 겁니다! 

---

현재 구현 사항이 단위 테스트라고 말씀해주셨는데요. 제 생각에는 `통합 테스트` 같아요.

먼저 스스로 단위 vs 통합 테스트에 대해 정의내려봐야 할 거 같아요.

간단하게 적어본다면

단위 테스트라면 단위와 범위를 작게, 외부 의존성을 적게 또는 없이 테스트합니다. 
다시 말해 외부 의존하고 있는 서비스(e.g. DB나 OAuth 서버 는 Mock)에 대한 범위를 끊거나, 만약 단위 테스트 범위가 서비스 레이어 테스트라면 repository 는 mock 처리 할 수 있습니다. 또는 순수 자바 객체에 대해 테스트할 수 있죠.

통합 테스트는 단위 테스트에서 체크하지 못하는 부분(외부 의존성), 좀 더 큰 범위를 테스트할 수 있습니다.
HTTP Request 요청 객체에 대한 validation 테스트나 DB와 통신하는 테스트를 합니다.

---

`@SpringBootTest`에 대해 생각해봐야 합니다.
해당 어노테이션은 Spring Boot의 모든 Context가 올라오기 때문에 매우 느린 테스트가 되고, 현재 TC(Test case)는 서비스 레이어만 테스트하는 것이 아닌 실제 repository까지 통신하고 있습니다. 위 정의를 따른다면 서비스-리포지토리 까지 테스트하는 통합테스트라고 생각됩니다.

---

spring 공식 문서에 적혀있는 내용을 참고하면 좋을 거 같습니다.  
해당 링크는 unit vs integration testing에 대해 설명하고 있습니다. 추가적으로 spring 에서 테스트할 때 context를 어디까지 사용야할지(Test Scope) 도 참고하면 좋을 거 같아요.

https://docs.spring.io/spring-framework/docs/current/reference/html/testing.html#testing-introduction
https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.testing

ku-kim commented on noriter/src/test/java/com/codewarts/noriter/article/unit/service/PlaygroundServiceTest.java:
```diff
@@ -0,0 +1,342 @@
+            .orElseThrow(() -> new GlobalNoriterException(ArticleExceptionType.ARTICLE_NOT_FOUND));
+
+        // then
+        assertThat(article.getId()).isEqualTo(articleId);
+        assertThat(article.getTitle()).isEqualTo(request.getTitle());
+        assertThat(article.getContent()).isEqualTo(request.getContent());
+        assertThat(article.getWriter().getId()).isEqualTo(writerId);
```
+a) 이 방식도 좋지만 중간에 하나 실패하면 아래 테스트는 작동하지 않아서요.! 
assertAll이나 SoftAssertions 방법도 있어요

ref

https://kukim.tistory.com/136 (예전에 작성했던 글인데 지금 보니 아쉬운게 있네요)
SoftAssertions 코드를 softly -> {} 람다 형식으로 쓸 수 있어요. 다른 글 참고해보시면 좋을 거 같습니다 !

PhilSoGooood replied:
오~ 이런게 있는줄 몰랐네요!! 🥹 참고해서 softAssertions 람다 형식으로 수정해보겠습니당

ku-kim commented on noriter/src/test/java/com/codewarts/noriter/article/unit/service/PlaygroundServiceTest.java:
```diff
@@ -0,0 +1,342 @@
+    @Test
+    void findDetailWithWrongArticleIdThenThrow() {
+        // expected
+        Assertions.assertThatThrownBy(() -> playgroundService.findDetail(9999L, writerId))
```
앗 여기는 Assertions가 포함되어있네요

추가적으로 9999L이 존재하지 않는 글이란거는 어떻게 아는걸까요?  테스트 코드만 보면 바로 알 수 있나욥?

PhilSoGooood replied:
그러면 Long.MAX_VALUE로 하면 될까요? 이게 음수인 경우는 validate에 걸리기때문에 가장 큰 단위의 Long 값 외에는 딱히 떠오르는 방법이 없네요...🤔

Assertions는 눈치 없이 들어가있었네용 삭제하겠습니당

ku-kim commented on noriter/src/test/java/com/codewarts/noriter/article/unit/service/PlaygroundServiceTest.java:
```diff
@@ -0,0 +1,342 @@
+    @DisplayName("자유게시판 글을 삭제한다.")
+    @Test
+    void delete() {
+        // given
+        PlaygroundCreateRequest createRequest = new PlaygroundCreateRequest("테스트 제목1", "테스트 내용1",
+            List.of("해시태그1", "해시태그2"));
+        Long articleId = playgroundService.create(createRequest, writerId);
+
+        // when
+        playgroundService.delete(articleId, writerId);
+        Article article = articleRepository.findById(articleId).get();
+
+        // then
+        Assertions.assertThat(article.isDeleted()).isTrue();
```
+a) 저 같은 경우에는 given, when, then을 주석으로 명시하지 않고 개행으로 구분하고 있습니다.
테스트코드 양이 많아질 경우에만 작성하고있어요.

PhilSoGooood replied:
어떤 방식을 말하는건지 찍먹이랑 자리톡 코드를 참고해도 잘 모르겠네요...🥸 조금 더 고민해보도록 하겠습니당 감사해요!

ku-kim reviewed (approved):
테스트를 시도해보셨군요 ~ 고생 많이하셨어요! 🤩

ku-kim commented:
시간 괜찮으실 때 최범균님 영상도 추천드려요.!

퓨어 자바 + tdd 연습
https://www.youtube.com/live/5PDYHNCcjYM?feature=share

테스트 시리즈 영상인데 1편부터 찾아서 슬쩍 감상 추천
https://www.youtube.com/watch?v=L_1UYlJyNuk

PhilSoGooood committed:
:recycle: PlaygroundService 테스트 리뷰 반영

- SoftAssertions 사용
- 기존 존재하지 않는 회원 및 글 조회시 9999L라고 하드코딩하던 것을 NON_EXIST_INDEX라고 상수 처리

Co-authored-by: rkolx <olx.rko.o@gmail.com>
